neural_nets/losses/losses.py

```diff
@@ -385,3 +385,66 @@ def grad(self, Y_fake, module, Y_real=None, gradInterp=None):
         else:
             raise ValueError("Unrecognized module: {}".format(module))
         return grad
+
+
+class Cosine(ObjectiveBase):
+    def __init__(self):
+        super().__init__()
+
+    def __call__(self, y, y_pred):
+        return self.loss(y, y_pred)
+
+    def __str__(self):
+        return "Cosine Proximity"
+
+    @staticmethod
+    def loss(y, y_pred):
+        """
+        Compute the Cosine distance between 1-D arrays.
+
+        The Cosine distance between `u` and `v`, is defined as
+        .. math::
+            1 - \\frac{u \\cdot v}
+                      {||u||_2 ||v||_2}.
+        where :math:`u \\cdot v` is the dot product of :math:`u` and
+        :math:`v`.
+
+        Parameters
+        ----------
+        y : numpy array of shape (n, m)
+
+        y_pred : numpy array of shape (n, m)
+
+
+        Returns
+        -------
+        loss : float
+
+        """
+        def l2_normalize(x, axis=-1):
+            y = np.max(np.sum(x ** 2, axis, keepdims=True), axis, keepdims=True)
+            return x / np.sqrt(y)
+
+        y_true = l2_normalize(y, axis=-1)
+        y_pred = l2_normalize(y_pred, axis=-1)
+        return 1. - np.sum(y_true * y_pred, axis=-1)
+
+    @staticmethod
+    def grad(y, y_pred, z, act_fn):
```

Comment:
Is there a sufficient way to do grad of cosine?

Reply:
Doing this from my phone, so please check for errors:

If `f(x, y) = (x @ y) / (norm(x) * norm(y))`, then we have

```
df/dy = sum(x) / (norm(x) * norm(y)) - (f(x, y) * y) / norm(y)
```
where `norm` is just the 2-norm

Reply:
Note that since cosine loss == negative cosine distance, you should multiply df/dy by -1